include/pybind11/numpy.h

```diff
@@ -1534,31 +1534,49 @@ struct vectorize_helper {
         // not wrapped in an array).
         if (size == 1 && ndim == 0) {
             PYBIND11_EXPAND_SIDE_EFFECTS(params[VIndex] = buffers[BIndex].ptr);
-            return cast(f(*reinterpret_cast<param_n_t<Index> *>(params[Index])...));
+            return apply_to_single_value(params, i_seq);
         }
 
-        array_t<Return> result;
-        if (trivial == broadcast_trivial::f_trivial) result = array_t<Return, array::f_style>(shape);
-        else result = array_t<Return>(shape);
-
-        if (size == 0) return std::move(result);
-
         /* Call the function */
         if (trivial == broadcast_trivial::non_trivial)
-            apply_broadcast(buffers, params, result, i_seq, vi_seq, bi_seq);
+            return apply_broadcast(buffers, params, trivial, shape, size, i_seq, vi_seq, bi_seq);
         else
-            apply_trivial(buffers, params, result.mutable_data(), size, i_seq, vi_seq, bi_seq);
+            return apply_trivial(buffers, params, trivial, shape, size, i_seq, vi_seq, bi_seq);
+    }
+
+    array_t<Return> create_array(broadcast_trivial trivial, const std::vector<ssize_t>& shape) {
+        if (trivial == broadcast_trivial::f_trivial)
+            return array_t<Return, array::f_style>(shape);
+        else
+            return array_t<Return>(shape);
+    }
+
+    template <bool flag=!is_noreturn, typename std::enable_if<flag>::type* =nullptr,
+              size_t... Index>
+    object apply_to_single_value(const std::array<void *, N> &params, const index_sequence<Index...>&) {
+        return cast(f(*reinterpret_cast<param_n_t<Index> *>(params[Index])...));
+    }
 
-        return std::move(result);
+    template <bool flag=is_noreturn, typename std::enable_if<flag>::type* =nullptr,
+              size_t... Index>
+    py::none apply_to_single_value(const std::array<void *, N> &params, const index_sequence<Index...>&) {
```

Comment:
Need to remove the `py::` namespaces, as the namespace alias is not defined by pybind11